### PR TITLE
Refactor dependency management logic for consistency and separation from UI

### PR DESCRIPTION
**Pull request by niezbop**
This makes sure that the dependency state is being kept tracked of in a consistent way, as well as keeping the lockfile udpated along update operations.

__Improvements:__

- Extract logic from `Uplift > Check Dependencies` and the `UpdateUtility` window to only keep display role, and move that logic to the UpliftManager in `GetDependenciesState()`.

- Use the `LogAggregator` in `Uplift > Check Dependencies` so that it doesn't flood the console.

__Fixes:__

- Previously mentionned refactoring made sure that the logic is consistent with the project requirements and the Update window will no longer allow to update to conflicting requirements.

- InstallPackage has been lightly modified to make sure that Update actions will keep the lockfile updated.

Both of these fixes address #5

### CODE REVIEW

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs [hunk 1]
```diff
@@ -79,13 +79,78 @@ public enum InstallStrategy {
...
+                        ref dependenciesState,
+                        definition,
+                        targets,
+                        any_installed
+                    );
+
+            return dependenciesState.ToArray();
+        }
+
+        public void AppendDependencyState(
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/UpliftManager.cs]: make non public?

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs [hunk 2]
```diff
@@ -79,13 +79,78 @@ public enum InstallStrategy {
...
             InstallPackages(targets);
         }
 
+        public DependencyState[] GetDependenciesState(bool refresh = false)
+        {
+            if (refresh) UpliftManager.ResetInstances();
+            Upbring upbring = Upbring.Instance();
+            PackageRepo[] targets = GetTargets(GetDependencySolver(), InstallStrategy.UPDATE_LOCKFILE);
+
+            bool any_installed =
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/UpliftManager.cs]: `anyInstalled`?

File: Assets/Plugins/Editor/Uplift/Windows/UpdateWindow.cs [hunk 3]
```diff
@@ -35,165 +35,103 @@ namespace Uplift.Windows
...
-                    manager.UpdatePackage(latestPackageRepo);
-
-                    AssetDatabase.Refresh();
-
-                    Repaint();
-                }
-                GUI.enabled = true;
+                if(!definition.Requirement.IsMetBy(installed.Version))
+                    EditorGUILayout.HelpBox(
+                        "The version of the package currently installed does not match the requirements of your project!",
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/Windows/UpdateWindow.cs]: maybe display the version and the requirements?

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs [hunk 4]
```diff
@@ -516,6 +581,31 @@ private void InstallPackage(Upset package, TemporaryDirectory td, DependencyDefi
...
+                        {
+                            found = true;
+                            break;
+                        }
+                    }
+                    if(found)
+                        snapshot.installableDependencies[index].Package = package;
+                    else
+                    {
+                        PackageRepo[] temp = snapshot.installableDependencies;
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/UpliftManager.cs]: Would this work?
```
snapshot.installableDependencies = Array.Resize<PackageRepo[]>(snapshot.installableDependencies, snapshot.installableDependencies.Length + 1);
snapshot.installableDependencies[snapshot.installableDependencies.Length] = = new PackageRepo { Package = package };
```

File: Assets/Plugins/Editor/Uplift/MenuItems.cs [hunk 5]
```diff
@@ -70,39 +71,50 @@ private static void EditUpfile()
...
+
+                    if(state.latest.Package.PackageVersion != state.bestMatch.Package.PackageVersion)
+                        message += string.Format(
+                            "\nNote: there is a more recent version of the package ({0} from {1}), but it doesn't match your requirement",
+                            state.latest.Package.PackageVersion,
+                            state.bestMatch.Repository.ToString()
+                        );
+
+                    if(state.transitive)
+                        message = "`--> " + string.Join("    \n", message.Split('\n'));
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/MenuItems.cs]: was the ` character expected?

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs [hunk 6]
```diff
@@ -79,13 +79,78 @@ public enum InstallStrategy {
...
+        }
+        
         public void InstallDependencies(InstallStrategy strategy = InstallStrategy.UPDATE_LOCKFILE, bool refresh = false)
         {
             if (refresh) UpliftManager.ResetInstances();
             PackageRepo[] targets = GetTargets(GetDependencySolver(), strategy);
             InstallPackages(targets);
         }
 
+        public DependencyState[] GetDependenciesState(bool refresh = false)
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/UpliftManager.cs]: Do we want to add the install strategy argument as in the above method's API? Adding it later would make them "out of order".

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs [hunk 7]
```diff
@@ -79,13 +79,78 @@ public enum InstallStrategy {
...
+                    );
+
+            return dependenciesState.ToArray();
+        }
+
+        public void AppendDependencyState(
+            ref List<DependencyState> dependenciesState,
+            DependencyDefinition definition,
+            PackageRepo[] targets,
+            bool any_installed,
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/UpliftManager.cs]: anyInstalled?

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs [hunk 8]
```diff
@@ -79,13 +79,78 @@ public enum InstallStrategy {
...
+        {
+            Upbring upbring = Upbring.Instance();
+
+            DependencyState state = new DependencyState
+            {
+                definition = definition,
+                latest = PackageList.Instance().GetLatestPackage(definition.Name)
+            };
+
+            state.bestMatch = targets.First(pr => pr.Package.PackageName == definition.Name);
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/UpliftManager.cs]: Given that we just refreshed, could that fail? I.e. `targets.First` return `null`?

File: Assets/Plugins/Editor/Uplift/MenuItems.cs (hunk 5 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/MenuItems.cs]: Yes, so this looks a bit like an arrow going down. I think it looks good, but we can remove it if you want

File: Assets/Plugins/Editor/Uplift/Windows/UpdateWindow.cs (hunk 3 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/Windows/UpdateWindow.cs]: I believe that this is not necessary as it is displayed just above:
capture d ecran 9

File: Assets/Plugins/Editor/Uplift/MenuItems.cs (hunk 5 above)
Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/MenuItems.cs]:
```
‎________$$$$
_______$$__$
_______$___$$
_______$___$$
_______$$___$$
________$____$$
________$$____$$$
_________$$_____$$
_________$$______$$
__________$_______$$
____$$$$$$$________$$
__$$$_______________$$$$$$
_$$____$$$$____________$$$
_$___$$$__$$$____________$$
_$$________$$$____________$
__$$____$$$$$$____________$
__$$$$$$$____$$___________$
__$$_______$$$$___________$
___$$$$$$$$$__$$_________$$
____$________$$$$_____$$$$
____$$____$$$$$$____$$$$$$
_____$$$$$$____$$__$$
_______$_____$$$_$$$
________$$$$$$$$$$
```

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs (hunk 6 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/UpliftManager.cs]: I don't think so: we want to be sure that the install strategy is `InstallStrategy.UPDATE_LOCKFILE` to be sure that we get all the best targets that could update the packages.

File: Assets/Plugins/Editor/Uplift/MenuItems.cs (hunk 5 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/MenuItems.cs]: This is a bit confusing though, do you approve, or do you think we should indeed remove it

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs (hunk 8 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/UpliftManager.cs]: I'm not sure. We go through the dependencies as does the targets, so there should not be an issue. We can still check for null though, but what behaviour should we have then?

File: Assets/Plugins/Editor/Uplift/MenuItems.cs (hunk 5 above)
Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/MenuItems.cs]: keep it

File: Assets/Plugins/Editor/Uplift/Windows/UpdateWindow.cs (hunk 3 above)
Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/Windows/UpdateWindow.cs]: Fine as is.

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs (hunk 8 above)
Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/UpliftManager.cs]: We keep it as is

**Pull request merged**